relentless/optimize/method.py

```diff
@@ -339,14 +469,38 @@ def optimize(self, objective):
         iter_num = 0
         cur_res = objective.compute()
         while not self.has_converged(cur_res) and iter_num < self.max_iter:
+            grad_y = cur_res.gradient
+            if np.isscalar(self.scale):
+                grad_y *= self.scale**2
+            else:
+                for x in self.scale:
+                    try:
+                        grad_y[x] *= self.scale[x]**2
+                    except KeyError:
+                        raise KeyError('''The scaling parameters cannot be defined
+                                          on a variable which is not in the objective
+                                          function.''')
+            update = self.descent_amount(grad_y)*grad_y
```

Comment:
Here is where we will need the second factor of the `scale`. So each element of `update` needs to have something like `self.descent_amount(grad_y)*grad_y[x]*scale[x]`. There is probably a way of handling things so you don't need an extra loop here.

FYI, we can't just stuff the scale^2 into grad_y because when we do FixedStep, we will need `descent_amount` to get the correct grad_y.

Reply:
Indeed, you do need to compute `grad_y = scale*grad_x` for `descent_amount`. The `KeyedArray` should allow you to do the update without a loop.

Reply:
Not sure this is possible, as the `scale` doesn't necessarily have all the design variables that `gradient` has, and we can't convert the input `scale` dictionary to a `KeyedArray` as the `SteepestDescent` isn't aware of the `ObjectiveFunction` design variables at the time of its construction...

Reply:
At the beginning of `optimize`, make a local `scale` variable with the right keys:
```py
scale = KeyedArray(dvars)
for x in dvars:
    if np.isscalar(self.scale):
        scale[x] = self.scale
    else:
        try:
            scale[x] = self.scale[x]
        except KeyError:
            scale[x] = 1.0
```